### PR TITLE
Optimize ResourceManager

### PR DESCRIPTION
**Pull request by lahma**
This starts to take shape, don't know how many bugs I've introduced but biggest offenders should be handled now.

``` ini
BenchmarkDotNet=v0.11.5, OS=Windows 10.0.18980
Intel Core i7-6820HQ CPU 2.70GHz (Skylake), 1 CPU, 8 logical and 4 physical cores
.NET Core SDK=3.0.100-preview9-014004
  [Host]     : .NET Core 3.0.0-preview9-19423-09 (CoreCLR 4.700.19.42102, CoreFX 4.700.19.42104), 64bit RyuJIT
  DefaultJob : .NET Core 3.0.0-preview9-19423-09 (CoreCLR 4.700.19.42102, CoreFX 4.700.19.42104), 64bit RyuJIT
```

## Before

|           Method |     Mean |     Error |    StdDev |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|----------------- |---------:|----------:|----------:|-------:|------:|------:|----------:|
| RenderStylesheet | 18.03 us | 0.0706 us | 0.0589 us | 3.6926 |     - |     - |  15.16 KB |

## After

|           Method |     Mean |     Error |    StdDev |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|----------------- |---------:|----------:|----------:|-------:|------:|------:|----------:|
| RenderStylesheet | 9.379 us | 0.0670 us | 0.0627 us | 1.9836 |     - |     - |   8.16 KB |

### CODE REVIEW

File: src/OrchardCore/OrchardCore.Mvc.Core/ShellFileVersionProvider.cs [hunk 1]
```diff
@@ -22,7 +22,7 @@ public class ShellFileVersionProvider : IFileVersionProvider
         private static readonly char[] QueryStringAndFragmentTokens = new[] { '?', '#' };
         private static readonly MemoryCache _sharedCache = new MemoryCache(new MemoryCacheOptions());
 
-        private readonly IEnumerable<IFileProvider> _fileProviders;
+        private readonly IFileProvider[] _fileProviders;
```
Comment (lahma) [inline src/OrchardCore/OrchardCore.Mvc.Core/ShellFileVersionProvider.cs]: foreach on array is as fast as it gets

File: src/OrchardCore/OrchardCore.ResourceManagement.Abstractions/RequireSettings.cs [hunk 2]
```diff
@@ -48,6 +50,7 @@ public bool HasAttributes
         /// <summary>
         /// The resource will be displayed in the head of the page
         /// </summary>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
```
Comment (lahma) [inline src/OrchardCore/OrchardCore.ResourceManagement.Abstractions/RequireSettings.cs]: all methods here are are quite small, it doesn't hurt to hint about inlining, especially when call site is Combine which calls all of them

File: src/OrchardCore/OrchardCore.ResourceManagement.Abstractions/ResourceDefinition.cs [hunk 3]
```diff
@@ -334,8 +335,13 @@ public override bool Equals(object obj)
 
         public override int GetHashCode()
         {
-            return (Name ?? "").GetHashCode() ^ (Type ?? "").GetHashCode();
+            return HashCode.Combine(Name, Type);
```
Comment (lahma) [inline src/OrchardCore/OrchardCore.ResourceManagement.Abstractions/ResourceDefinition.cs]: The new HashCode class should pretty much used always, https://github.com/dotnet/corefx/issues/14354

File: src/OrchardCore/OrchardCore.ResourceManagement.Abstractions/ResourceManifest.cs [hunk 4]
```diff
@@ -6,7 +6,7 @@ namespace OrchardCore.ResourceManagement
     public class ResourceManifest
     {
         private string _basePath = "";
-        private readonly IDictionary<string, IDictionary<string, IList<ResourceDefinition>>> _resources = new Dictionary<string, IDictionary<string, IList<ResourceDefinition>>>(StringComparer.OrdinalIgnoreCase);
+        private readonly Dictionary<string, IDictionary<string, IList<ResourceDefinition>>> _resources = new Dictionary<string, IDictionary<string, IList<ResourceDefinition>>>(StringComparer.OrdinalIgnoreCase);
```
Comment (lahma) [inline src/OrchardCore/OrchardCore.ResourceManagement.Abstractions/ResourceManifest.cs]: really would like to have concrete types here, but would affect public api

File: src/OrchardCore/OrchardCore.ResourceManagement.Abstractions/ResourceDefinition.cs [hunk 5]
```diff
@@ -98,7 +99,7 @@ public ResourceDefinition SetUrl(string url, string urlDebug)
         {
             if (String.IsNullOrEmpty(url))
             {
-                throw new ArgumentNullException("url");
```
Comment (lahma) [inline src/OrchardCore/OrchardCore.ResourceManagement.Abstractions/ResourceDefinition.cs]: here we have many small methods that could not be inlined when they had `throw`, now letting JIT to decide

File: src/OrchardCore/OrchardCore/Environment/Shell/RunningShellTable.cs [hunk 6]
```diff
@@ -100,7 +100,7 @@ public ShellSettings Match(string host, string appRelativePath, bool fallbackToD
         private bool TryMatchInternal(string host, string appRelativePath, bool fallbackToDefault, out ShellSettings result)
         {
             // We can skip some checks for port matching if we know the host doesn't contain one
-            var hasPort = host.IndexOf(':') != -1;
+            var hasPort = host.Contains(':');
```
Comment (lahma) [inline src/OrchardCore/OrchardCore/Environment/Shell/RunningShellTable.cs]: Like @jtkech correctly pointed out it's fine to use Contains with netcoreapp3.0/nestandard2.1, it doesn't use LINQ

File: src/OrchardCore/OrchardCore.ResourceManagement/ResourceManager.cs [hunk 7]
```diff
@@ -13,7 +15,7 @@ namespace OrchardCore.ResourceManagement
     public class ResourceManager : IResourceManager
     {
         private readonly Dictionary<ResourceTypeName, RequireSettings> _required = new Dictionary<ResourceTypeName, RequireSettings>();
-        private readonly Dictionary<string, IList<ResourceRequiredContext>> _builtResources;
+        private readonly Dictionary<string, ResourceRequiredContext[]> _builtResources;
```
Comment (lahma) [inline src/OrchardCore/OrchardCore.ResourceManagement/ResourceManager.cs]: preferring concrete types, here even array was possible, which is the fastest

File: src/OrchardCore/OrchardCore.ResourceManagement/ResourceManager.cs [hunk 8]
```diff
@@ -173,52 +171,14 @@ private ResourceDefinition FindResource(RequireSettings settings, bool resolveIn
             // using the action.
             var name = settings.Name ?? "";
             var type = settings.Type;
-            ResourceDefinition resource;
```
Comment (lahma) [inline src/OrchardCore/OrchardCore.ResourceManagement/ResourceManager.cs]: I rewrote this without LINQ, code probably is even shorter now, this is potential bug source with setting version etc matching

File: src/OrchardCore/OrchardCore.ResourceManagement/ResourceManager.cs [hunk 9]
```diff
@@ -308,46 +314,53 @@ private bool ResolveInlineDefinitions(string resourceType)
...
+                {
+                    yield return value;
+                }
             }
-
-            return _links.AsReadOnly();
         }
 
-        public IEnumerable<MetaEntry> GetRegisteredMetas()
+        public IEnumerable<LinkEntry> GetRegisteredLinks() => DoGetRegisteredLinks();
```
Comment (lahma) [inline src/OrchardCore/OrchardCore.ResourceManagement/ResourceManager.cs]: here's public API + DoAPI, Do prefix uses concrete types and is faster when called inside the class, public API exposes IEnumerable which is slowest to traverse (.NET Core has made huge progress here checking the type behind IEnumerable, but still).

File: src/OrchardCore/OrchardCore.ResourceManagement/ResourceManager.cs [hunk 10]
```diff
@@ -358,44 +371,70 @@ public IEnumerable<ResourceRequiredContext> GetRequiredResources(string resource
...
-                                   let version = segments.Length > 1 ? segments[1] : null
-                                   select FindResource(new RequireSettings { Type = resource.Type, Name = name, Version = version });
-                foreach (var dependency in dependencies)
+                // share search instance
+                var tempSettings = new RequireSettings();
+
+                for (var i = 0; i < resource.Dependencies.Count; i++)
                 {
+                    var d = resource.Dependencies[i];
+                    var idx = d.IndexOf(':');
```
Comment (lahma) [inline src/OrchardCore/OrchardCore.ResourceManagement/ResourceManager.cs]: IndexOf is cheaper than Split now that we know to have one or two items

File: src/OrchardCore/OrchardCore.ResourceManagement/ResourceManager.cs [hunk 11]
```diff
@@ -618,29 +672,15 @@ public ResourceTypeName(string resourceType, string resourceName)
...
-            {
-                var sb = new StringBuilder();
-                sb.Append("(");
-                sb.Append(Type);
-                sb.Append(", ");
-                sb.Append(Name);
-                sb.Append(")");
-                return sb.ToString();
-            }
+            public override string ToString() => "(" + Type + ", " + Name + ")";
```
Comment (lahma) [inline src/OrchardCore/OrchardCore.ResourceManagement/ResourceManager.cs]: StringBuilder can be an overkill too, concatenating < 10 strings usually is faster than doing the StringBuilder dance

File: src/OrchardCore/OrchardCore.ResourceManagement.Abstractions/ResourceDefinition.cs [hunk 12]
```diff
@@ -127,7 +128,7 @@ public ResourceDefinition SetCdnIntegrity(string cdnIntegrity, string cdnDebugIn
         {
             if (String.IsNullOrEmpty(cdnIntegrity))
             {
-                throw new ArgumentNullException("cdnUrl");
+                ThrowArgumentNullException("cdnUrl");
```
Comment (sebastienros) [inline src/OrchardCore/OrchardCore.ResourceManagement.Abstractions/ResourceDefinition.cs]: Is that a mistake?
```suggestion
                ThrowArgumentNullException(nameof(cdnIntegrity));
```

Comment (lahma) [reply src/OrchardCore/OrchardCore.ResourceManagement.Abstractions/ResourceDefinition.cs]: Good catch!

**Comment by sebastienros** [comment]
Should I merge or are you doing more on this branch?

**Comment by lahma** [comment]
I think biggest offenders have been taken care of, so merge is fine. There's little test coverage so hopefully I didn't break it badly...

**Pull request merged**